pkg/v1/validate/image.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-containerregistry/pkg/v1"
+)
+
+// Image valides that img does not violate any invariants on the image format.
+func Image(img v1.Image) error {
+	if err := validateLayers(img); err != nil {
+		return fmt.Errorf("validating layers: %v", err)
+	}
+
+	if err := validateConfig(img); err != nil {
+		return fmt.Errorf("validating config: %v", err)
+	}
+
+	if err := validateManifest(img); err != nil {
+		return fmt.Errorf("validating manifest: %v", err)
+	}
+
+	return nil
+}
+
+func validateConfig(img v1.Image) error {
+	cn, err := img.ConfigName()
+	if err != nil {
+		return err
+	}
+
+	rc, err := img.RawConfigFile()
+	if err != nil {
+		return err
+	}
+
+	hash, size, err := v1.SHA256(bytes.NewReader(rc))
+	if err != nil {
+		return err
+	}
+
+	m, err := img.Manifest()
+	if err != nil {
+		return err
+	}
+
+	cf, err := img.ConfigFile()
+	if err != nil {
+		return err
+	}
+
+	pcf, err := v1.ParseConfigFile(bytes.NewReader(rc))
+	if err != nil {
+		return err
+	}
+
+	if cn != hash {
+		return fmt.Errorf("mismatched config digest: ConfigName()=%s, SHA256(RawConfigFile())=%s", cn, hash)
```

Comment:
What do you think about collecting validation errors and reporting all of them at once at the end? This can help make debugging faster since you can see "the digest was wrong" (not that helpful on its own) alongside "the config content was wrong" ("oh _that's_ why the digest was wrong!").

Reply:
Done.